File: Scripts/realTimehandler.js
var realtimehandler;

$(function () {
    realtimehandler = $.connection.realTimeHandler;
    $.connection.hub.start();
    //添加消息
    realtimehandler.client.addMessage = function(m) {
        $("#userdiv").html($("#userdiv").html()+"<br />"+m);
    };
    //添加用户
    realtimehandler.client.addUser = function (m) {
        for (i = 0; i < m.length; i++) {
            $("#userdiv").html($("#userdiv").html() + "<br />" + m[i]);
        }
    };
    //接收消息
    realtimehandler.client.receiveMsg = function (connectionid, userid, s) {
        var clr = layersColors[layersColors.indexOf(connectionid) + 1];
        $("#messagediv").html($("#messagediv").html() + "<br /><font color='"+clr+"'>" + userid + "</font>:" + s);
    };
    //添加图层
    realtimehandler.client.addLayer = function (connectionid, clr) {
        addlayer(connectionid, clr);
        if ($.connection.hub.id == connectionid)
        {
            setcolor($.connection.hub.id,clr);
        }
    }
    //刷新用户
    realtimehandler.client.refreshUser = function refreshUser(m,c) {     
        var html="<table>";
        for (i = 0; i < m.length; i++) {
            html += "<tr><td><font  color='" + c[i] + "'>" + m[i] + "</font></td></tr>";
        }
        html += "</table>";
        $("#userdiv").html(html);
    };
    //添加面
    realtimehandler.client.addPolygon = function (id, path) {
        addPolygon(id, path);
    }
    //添加线
    realtimehandler.client.addPolyline = function (id, path) {
        addPolyline(id, path);
    }
    //添加点
    realtimehandler.client.addPoint = function (id, x, y) {
        addPoint(id, x, y);
    }
    //添加Extent
    realtimehandler.client.addExtent = function (id,json) {
        addExtent(id,json);
    }
    //清除图层
    realtimehandler.client.clear = function (id)
    {
        clear(id);
    }
    //设置用户
    realtimehandler.client.setuser = function (user)
    {
        $("#txtuser").val(user);
    }
    //信息发送按钮事件
    $("#btnsend").click(function () {
        realtimehandler.server.sendMessage($("#txtmessage").val());
        $("#txtmessage").val("");
    });
    //用户名更新
    $("#btnupdate").click(function () {
        realtimehandler.server.updateuser($("#txtuser").val());
    });

})

File: App_Code/RealTimeHandler.cs
using System;
using System.Collections.Generic;
using System.Web;
using System.Threading.Tasks;
namespace SingnalR
{
    /// <summary>
    /// RealTimeHandler 的摘要说明
    /// </summary>
    public class RealTimeHandler : Microsoft.AspNet.SignalR.Hub
    {
        public static Dictionary<String, String> _usernames = new Dictionary<string, string>();
        public static Dictionary<String, String> _legends = new Dictionary<string, string>();
        public static int _index = 0;
        public static List<string> _Colors = new List<string>() { "red", "blue", "green", "black", "grey", "purple", "fuchsia", "silver", "lime", "olive", "yellow", "navy", "teal", "aqua", "maroon" };

        public RealTimeHandler()
        {
            //
            // TODO: 在此处添加构造函数逻辑
            //
        }
        /// <summary>
        /// 用户连接时触发
        /// </summary>
        /// <returns></returns>
        public override System.Threading.Tasks.Task OnConnected()
        {
            String connectionid = Context.ConnectionId;

            foreach (String key in _legends.Keys)
            {
                Clients.Caller.addLayer(key, _legends[key]);
            }
            String username="user" + (++_index);
            _usernames.Add(connectionid, username);
            Clients.Caller.setuser(username);
            String color;
            lock (_Colors)
            {
                color = _Colors[0];
                _legends.Add(connectionid, _Colors[0]);
                _Colors.RemoveAt(0);
            }
            Clients.All.addLayer(connectionid, color);

            refreshusers();
            return base.OnConnected();
        }
        /// <summary>
        /// 更新客户端用户列表
        /// </summary>
        private void refreshusers()
        {
            String[] users = new String[_usernames.Count];
            String[] legends = new String[_usernames.Count];
            String[] keys = new String[_usernames.Count];
            _usernames.Keys.CopyTo(keys, 0);
            for (int i = 0; i < keys.Length; i++)
            {
                users[i] = _usernames[keys[i]];
                legends[i] = _legends[keys[i]];
            }

            Clients.All.refreshUser(users, legends);
        }
        /// <summary>
        /// 用户断开时触发
        /// </summary>
        /// <returns></returns>
        public override Task OnDisconnected()
        {
            _usernames.Remove(Context.ConnectionId);
            _Colors.Add(_legends[Context.ConnectionId]);
            _legends.Remove(Context.ConnectionId);
            refreshusers();

            return base.OnDisconnected();
        }
        /// <summary>
        /// 发送消息
        /// </summary>
        /// <param name="str"></param>
        public void SendMessage(String str)
        {
            Clients.All.receiveMsg(Context.ConnectionId,_usernames[Context.ConnectionId], str);
        }
        /// <summary>
        /// 绘制线
        /// </summary>
        /// <param name="paths"></param>
        public void AddPolyline(double[][][] paths)
        {
            Clients.All.addPolyline(Context.ConnectionId, paths);
        }
        /// <summary>
        /// 绘制Extent
        /// </summary>
        /// <param name="json"></param>
        public void AddExtent(String json)
        {
            Clients.All.addExtent(Context.ConnectionId, json);
        }
        /// <summary>
        /// 绘制面
        /// </summary>
        /// <param name="paths"></param>
        public void AddPolygon(double[][][] paths)
        {
            Clients.All.addPolygon(Context.ConnectionId, paths);

        }
        /// <summary>
        /// 绘制点
        /// </summary>
        /// <param name="x"></param>
        /// <param name="y"></param>
        public void AddPoint(double x, double y)
        {
            Clients.All.addPoint(Context.ConnectionId, x, y);
        }
        /// <summary>
        /// 清除图层
        /// </summary>
        public void Clear()
        {
            Clients.All.clear(Context.ConnectionId);
        }
        /// <summary>
        /// 更新用户名
        /// </summary>
        /// <param name="user"></param>
        public void updateuser(String user) {
            _usernames[Context.ConnectionId] = user;
            refreshusers();
        }
    }
}